jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java

```diff
@@ -0,0 +1,210 @@
+    }
+
+    private void printState(int state, int n){
+        System.out.printf("State %d: ", state);
+        for (int i = 0; i < n; i++) {
+            if (((state >> i) & 1) > 0){
+                System.out.printf("%d ", i);
+            }
+        }
+        System.out.println();
+    }
+
+    private double memo(int previousNode, int state, double[][] C, double[][] W){
+        // have we seen this state before?
+        if (C[previousNode][state] != Double.MIN_VALUE)
+            return C[previousNode][state];
+
+        // no cycle has been found yet
+        double totalCost = Double.MAX_VALUE;
+
+        // check if all nodes have been visited (i.e. state + 1 == 2^n)
+        if (state == (1 << W.length) - 1){
+            // check if there is a return edge we can use
+            if (W[previousNode][0] != Double.MAX_VALUE)
+                totalCost = W[previousNode][0];
+            else
+                totalCost = Double.MAX_VALUE;
+        }
+        else{
+            // try to find the 'best' next (i.e. unvisited and adjacent to previousNode) node in the tour
+            for (int i = 0; i < W.length; i++) {
+                if (((state >> i) & 1) == 0 && W[previousNode][i] != Double.MAX_VALUE){
+                    totalCost = Math.min(totalCost, W[previousNode][i] + memo(i, state ^ (1 << i), C, W));
+                }
+            }
+        }
+
+        return C[previousNode][state] = totalCost;
+    }
+
+    /**
+     * Computes a minimum-cost Hamiltonian tour.
+     *
+     * @param graph the input graph
+     * @return a minimum-cost tour if one exists, null otherwise
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        final int n = graph.vertexSet().size(); // number of nodes
+
+        if (n == 0) {
+            throw new IllegalArgumentException("Graph contains no vertices");
+        }
+
+        if (n == 1){
+            V startNode = graph.vertexSet().iterator().next();
+            return new GraphWalk<>(graph, startNode, startNode, Collections.singletonList(startNode), null, 0);
+        }
+
+        // W[u, v] = the cost of the minimum weight between u and v
+        double[][] W = new double[n][n];
+        for (int i = 0; i < n; i++) {
+            Arrays.fill(W[i], Double.MAX_VALUE);
+        }
+
+        /*
+         * Normalize the graph
+         *   map each vertex to an integer (using a HashMap)
+         *   keep the reverse mapping  (using an ArrayList)
+         */
+        Map<V, Integer> mapToInt = new HashMap<>();
+        List<V> mapToV = new ArrayList<>();
+        int newNode = 0;
+        for (E e: graph.edgeSet()){
+            V source = graph.getEdgeSource(e);
+            V target = graph.getEdgeTarget(e);
+
+            // map 'source' if no mapping exists
+            if (!mapToInt.containsKey(source)){
+                mapToInt.put(source, newNode);
```

Comment:
can be substituted by mapToIn.put(source,mapToInt.size())

Reply:
I am not sure I understand what you mean. mapToIn.put(source,mapToInt.size()) used instead of mapToIn.put(source, newNode) is not an improvement (and you can't use .putIfAbsent).

Reply:
The only difference is, that you don't need your `newNode` variable which gets incremented each time you add a node to the map (newNode is simply a counter which tracks the number of items in your map). This would make your code more concise (gets rid of 3 lines). Indeed not a 'performance' improvement.

Reply:
I have updated the code.